src/account/sync/mod.rs

```diff
@@ -1099,70 +1104,116 @@ impl SyncedAccount {
+                }
+                LEDGER_MAX_IN_OUTPUTS - output_amount
+            }
+            _ => {
+                if output_amount >= INPUT_OUTPUT_COUNT_MAX {
+                    return Err(crate::Error::TooManyOutputs(output_amount, INPUT_OUTPUT_COUNT_MAX));
+                }
+                INPUT_OUTPUT_COUNT_MAX
+            }
+        };
+
+        let mut available_inputs: Vec<input_selection::Input> = Vec::new();
+        for address_input in available_outputs {
+            let filtered: Vec<AddressOutput> = address_input.clone()
+                .outputs
+                .clone()
+                .into_iter()
+                .filter(|output| {
+                    (!transfer_obj.outputs.iter().any(|o| &o.address == output.address())
+                && *output.amount() > 0
+                && !locked_outputs.iter().any(|output|&output.transaction_id == output.transaction_id() && &output.index == output.index())
+                // we allow an input equal to a deposit address only if it has balance <= transfer amount, so there
+                // can't be a remainder value with this address as input alone
+                    || transfer_obj.outputs.iter().any(|o| &o.address == output.address())
+                        && *output.amount() <= transfer_obj.amount())
+                        && *output.amount() > 0
+                        && !locked_outputs.iter().any(|output| {
+                            &output.transaction_id == output.transaction_id() && &output.index == output.index()
+                        })
+                }).collect();
+            for output in filtered {
+                available_inputs.push(input_selection::Input {
+                    internal: address_input.internal,
+                    output: output.clone(),
+                });
+            }
         }
 
-        locked_addresses.extend(
-            selected_addresses
-                .iter()
-                .map(|a| a.address.clone())
-                .collect::<Vec<AddressWrapper>>(),
-        );
+        let selected_outputs = input_selection::select_input(transfer_obj.amount(), available_inputs, max_inputs)?;
+        locked_outputs.extend(selected_outputs.iter().map(|input| input.output.clone()));
+
+        let inputs_amount = selected_outputs.iter().fold(0, |acc, a| acc + a.output.amount);
+        let has_remainder = inputs_amount > transfer_obj.amount();
 
         let remainder = if has_remainder {
-            selected_addresses.last().cloned()
+            let input_for_remainder = selected_outputs
+                .iter()
+                // We filter the output addresses, but since we checked that this address balance <=
+                // transfer_obj.amount.get() we need to have another input address
+                .filter(|input| !transfer_obj.outputs.iter().any(|o| o.address == input.output.address))
+                .collect::<Vec<&input_selection::Input>>()
+                .last()
+                .cloned()
```

Comment:
Do we need the second cloned()?

Reply:
Yes, because last() returns `Option<&T>`